pkg/util/wsstream/conn.go

```diff
@@ -123,21 +125,30 @@ func handshake(config *websocket.Config, req *http.Request, allowed []string) er
 // Conn supports sending multiple binary channels over a websocket connection.
 // Supports only the "channel.k8s.io" subprotocol.
 type Conn struct {
-	channels []*websocketChannel
-	codec    codecType
-	ready    chan struct{}
-	ws       *websocket.Conn
-	timeout  time.Duration
+	channels           []*websocketChannel
+	codec              codecType
+	ready              chan struct{}
+	ws                 *websocket.Conn
+	timeout            time.Duration
+	supportedProtocols []string
+	protocolPrefix     string
 }
 
 // NewConn creates a WebSocket connection that supports a set of channels. Channels begin each
 // web socket message with a single byte indicating the channel number (0-N). 255 is reserved for
 // future use. The channel types for each channel are passed as an array, supporting the different
 // duplex modes. Read and Write refer to whether the channel can be used as a Reader or Writer.
-func NewConn(channels ...ChannelType) *Conn {
+//
+// One of the base64 ("base64.channel.k8s.io") and raw ("channel.k8s.io") protocol is negotiated
+// during initial handshake. Given protocolPrefixes (being DNS 1123 subdomains or empty string)
+// are provided all combinations of those and base64+raw are nogatiated, e.g. "", "v1", "v2" leads
+// to the protocols "base64.channel.k8s.io", "channel.k8s.io", "v1.base64.channel.k8s.io",
+// "v1.channel.k8s.io", "v2.base64.channel.k8s.io", "v2.channel.k8s.io".
+func NewConn(protocolPrefixes []string, channels ...ChannelType) *Conn {
```

Comment:
New conn doesn't really know anything about protocols - if you want to make it protocol aware, you should be passing in something that abstracts protocols, rather than make NewConn protocol aware.